### PR TITLE
NPE when listing sessions in UI if associated user is gone

### LINKED ISSUE #28801: NPE when listing sessions in UI if associated user is gone
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

When you remove a user federation provider or an imported user via a user federation and there are user sessions in the cache or DB, UI is not able to list sessions due to NPE.
Related issues https://github.com/keycloak/keycloak/issues/22428, https://github.com/keycloak/keycloak/issues/20443, https://github.com/keycloak/keycloak/issues/20170

### Version

24.0.2

### Regression

- [ ] The issue is a regression

### Expected behavior

To show sessions if the are orphaned sessions in the cache or DB.

### Actual behavior

NPE and UI error is shown

### How to Reproduce?

1. create federated user and sessions for the user
2. remove user federation provider
3. click on sessions tab

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by martin-kanis**
Closes #28801

This PR fixes the NPE when listing orphaned sessions (without associated user) in UI. It proactively removes the the orphaned session when detected.

### CODE REVIEW

**Review by pedroigor — COMMENTED** [review]
For now, it works fine so we don't have errors because of orphan sessions. Pagination might be eventually broken but better than before.

The fact we are lazily removing orphan sessions should also help with the original problem and reduce the chances of broken pagination.

I'm not 100% sure how far executing delete statements (e.g.: database with a lot of offline sessions) during session lookup will impact session look-up and authentication flows but let's try it out and get feedback.

Not sure if filtering orphan sessions when executing offline session queries will perform well when you have a huge amount of users and sessions.

@ahus1 @rmartinc @mposolda wdyt?

**Comment by ahus1** [comment]
There are the usual trade-offs: 

* When deleting the items while reading, you are self-correcting the data. As you are fetching only maxresult from the database, it is about to delete only maxresult (one page), and the result could still be empty - which is not perfect. The caller would see empty pages, which is confusing where they would assume they have reached the end of the pagination list. 
* When multiple people are reading at the same time, they would be trying delete items concurrently, which lead to problems in other places before when Hibernate throws exceptions. 

Once any skipping is involved, the pagination based on an index doesn't work well. I hope we would eventually move on to keyset pagination.

I usually lean towards skipping the entries and not deleting them when reading to avoid exceptions being thrown by Hibernate/the database. 

So to summarize: I'd add the skipping as done in the code, but not delete items on the fly, but wouldn't mind too much if you would keep it. 

With the persistent sessions, you could file an issue to improve that eventually (enqueue any invalid session we find for async deletion, or a background worker that does some cleanup)

**Comment by pedroigor** [comment]
> So to summarize: I'd add the skipping as done in the code, but not delete items on the fly, but wouldn't mind too much if you would keep it.

I'm 50/50 and more inclined to keep the deletion of items on the fly. My assumption is that should be a transitory state and it should not impact too much the majority of deployments. It should also "self-heal" as you remove stale data on-demand.

> With the persistent sessions, you could file an issue to improve that eventually (enqueue any invalid session we find for async deletion, or a background worker that does some cleanup)

Sounds a interesting and perhaps a middle ground to introduce a "controlled" mechanism for dealing with these sessions. I'll create one.

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@ahus1 Created https://github.com/keycloak/keycloak/issues/28813.
